### ISSUE TITLE
Flex output is trying to create ID index on tables without IDs

### ISSUE THREAD

**Issue by rouen-sk** [openstreetmap/osm2pgsql#1507]
Hey,
For my use case I need tables `way_nodes` and `relation_members`, similar to those created by osmosis tool. 
It is no problem to create them in osm2pgsql:

```
tables.way_nodes = osm2pgsql.define_table({
	name = 'way_nodes',
	columns = {
        { column = 'way_id', type = 'bigint' },
        { column = 'node_id', type = 'bigint' },
        { column = 'sequence_id', type = 'integer' }
	}
})

function osm2pgsql.process_way(object)
    ...
    for i,v in ipairs(object.nodes) do 
        tables.way_nodes:add_row({
          way_id = object.id,
          node_id = v,
          sequence_id = i
        })
    end		
    ...
end
```
And it works fine. But the import ultimately fails, because osm2pgqsl is trying to create ID index for these tables and creates invalid SQL statement for that (index on no columns):
```
2021-05-19 09:00:18  osm2pgsql version 1.4.2
2021-05-19 09:00:18  Database version: 12.6 (Ubuntu 12.6-1.pgdg18.04+1)
2021-05-19 09:00:18  PostGIS version: 3.1
2021-05-19 09:00:18  Node-cache: cache=4096MB, maxblocks=65536*65536, allocation method=3
2021-05-19 09:00:18  WARNING: Table 'relation_members' doesn't have an 'ids' column. Updates and expire will not work!
2021-05-19 09:00:56  Reading input files done in 38s.
2021-05-19 09:00:56    Processed 24073351 nodes in 8s - 3009k/s
2021-05-19 09:00:56    Processed 3298258 ways in 28s - 118k/s
2021-05-19 09:00:56    Processed 38094 relations in 2s - 19k/s
2021-05-19 09:00:57  Reprocess marked ways (stage 2)...
2021-05-19 09:00:57  Creating id index on table 'ways'...
2021-05-19 09:00:57  Creating id index on table 'relation_members'...
2021-05-19 09:00:57  node cache: stored: 24073351(100.00%), storage efficiency: 51.16% (dense blocks: 520, sparse nodes: 21395955), hit rate: 100.00%
2021-05-19 09:00:58  ERROR: Database error: ERROR:  syntax error at or near ")"
LINE 1: CREATE INDEX ON "public"."relation_members" USING BTREE ()
```

Is there a way to do one of these things?
* Define ID with multiple columns for output table? (for example `way_id` and `sequence_id` are composite key for `way_nodes`)
* Make osm2pgsql not creating indexes for these tables without IDs (and I will create them manually afterwards)

**Comment by joto** [comment]
osm2pgsql should not try to create an id index on a table that doesn't have an id. If that's the case that is a bug.

Your config file above is creating a table `way_nodes`, but the log file you post is failing on table `relation_members`. That's a bit confusing.

**Comment by rouen-sk** [comment]
Yeah, sorry for confusion, I have commented out `way_nodes` to try if `relation_members` have the same issue, and I've attached that output by mistake, but the error is the same for `way_nodes`. 

Now I have noticed another detail - I've tried to simplify my lua script to isolate the case, but if I leave out some other tables, the issue with index on `way_nodes` does not appear. It seems to me that things go wrong when I use "second stage" (by implementing `function osm2pgsql.select_relation_members(relation)` )

**Comment by joto** [comment]
We probably shouldn't try to create an index on ids on a table that doesn't have them. But with two-stage processing you really need that index, otherwise it will become horribly slow.

In your case I don't see why you can't have an id colum, though. Just create the `way_nodes` tables as a normal way table and the `relation_members` table as a normal relations table. They'll have the id that osm2pgsql needs. Then add your own composite index on `(way_id, node_ref)`. This will add some overhead due to the two indexes, but it gives osm2pgsql what it needs and you what you need. If you do a non-slim import or use `--drop`, you don't need the index any more after the import and you can drop it (or maybe its not generated at all, I am not sure at the moment).

**Comment by rouen-sk** [comment]
Huh, I did not think of it that way, I somehow assumed way tables (with way ID) must have unique ways in them, etc. But that is not the case clearly. 

I will try that. 

Meanwhile, I have patched my local version to disable ID index creation and deleting (as part of re-processing) for tables without ID columns - I have added `&& table.table().has_id_column()`

to https://github.com/openstreetmap/osm2pgsql/blob/f23e6ab522646ff47a6a786d8870d54e7b9c904a/src/output-flex.cpp#L1725
and
https://github.com/openstreetmap/osm2pgsql/blob/f23e6ab522646ff47a6a786d8870d54e7b9c904a/src/output-flex.cpp#L1500

And it works :) 
Your solution is probably the correct one for my use case, but my patch is probably still needed for scenarios where you really need  tables without IDs.
